examples/react-navigation/README.md

```diff
@@ -0,0 +1,5 @@
+# RNTL example app for React Navigation
+
+This example is shows how to use React Native Testing Library when testing apps using React Navigation.
```

Comment:
btw, i think it would be interesting to have a guide showing the 2 ways of testing navigation (with or without mocking it) wdyt? (and showing the pros and cons of each approach)

Reply:
@MattAgn what do you mean by mocking it?

Reply:
you render your component without the Navigation around it, then you mock `useNavigation` and you check the `navigate` function from useNavigation is called with the proper argument

Reply:
I think this is what majority of users are doing, not sure if it worth providing example for, since you mock various `useNavigation` props in different use-cases. Also I would like to encourage our users to leverage the full power of RNTL by building integration tests using the real libraries rather than doing shallow, hook-level, mocking.

What might be worth doing is showing how user can render a single screen using something like custom `renderScreen(ui, navigationParams)` which would leverage real `NavigatorContainer` + e.g. `StackNavigator`. Wdyt @MattAgn?

Reply:
What would renderScreen look like ? I'm not a huge fan of creating navigators in tests. Ideally you want to test your production code, writing navigators in tests doesn't test the navigator and also adds more code that needs to be maintained.  What I usually do when I want to test some flow that involves several pages is render the navigator containing the pages I need (can be the whole app or just a StackNavigator) and wrap with a NavigatorContainer if needed. That way I don't have to use any setup function or create false navigators that are equivalent to mocks, wdyt about this approach @mdjastrzebski?

Reply:
I've added that section, pls check the latest `test-utils.js`: `renderScreenWithParams`. 

re testing navigators vs screen: this PR contains both approaches `renderNavigator` renders passed JSX and just wraps it in `NavigatorContainer` while `renderScreenWithParams` allows you to pass route params to a given screen without using mocks, but using test setup involving `NavigatorContainer`, and simple `StackNavigator`. The goal here is to test single screen with different route params. Which might be difficult to do if that screen is buried deep in true navigators.

I think both approaches have their pros and cons, basically easy of test setup vs how well the test conditions resemble real life app. But I think both are still better than mocking `useNavigation` since they both work with the real React Navigation.

@pierrezimmermannbam let me know if that make sense to you or would you rather propose some other solution for single screen testing.

Reply:
I'm not sure I entirely understand your approach, how do you use renderScreenWithParams ? If react-navigation is not mocked then calls to navigate will throw if the screen we want to navigate to isn't mounted no ? Is it intended to test route params ?

Reply:
Yes, `navigate` calls will throw unless mocked e.g. with `jest.spyOn(...).mockImplementation(...)`. The main purpose of this kind of tests is to test a single screen that uses route params.

Reply:
It's very a interesting approach ! How do you manage to mock the navigation if it's passed down to the screen as a prop ?

On projects I've been working on so far we're using a different approach, namely because we mock globally @react-navigation/native in our tests and only unmock it for tests involving multiple pages. So when we have to test route params we basically render our component and pass the route prop directly to it

```tsx
render(<Page route={getMockRoute(params)} navigation={mockNavigation} />);
```

Where getMockRoute is a test util that returns an object that ressembles what route must look like with the provided params. This allows to easily test calls to mockNavigation methods.

Reply:
@pierrezimmermannbam I wonder is your approach working with components using `useNavigation` to get route params?

Reply:
I think you are right that in my `renderScreenWithParams` approach it's actually not possible to handle `navigation` mocking for `navigation` passed as a prop.

Reply:
No @pierrezimmermannbam approach won't handle calls to useNavigation since his Page is not wrapped with a NavigationContainer, providing the context used by `useNavigation`. I feel like each of your approach works for one scenario but not the other

Reply:
Imo, if we have to make a compromise anyway on how much our tests are realistic and can handle different scenarios, we might as well go all the way with @pierrezimmermannbam's approach which will be more performant wdyt?

Reply:
It does work with useNavigation since it's globally mocked but if components use useRoute to retrieve route params than it should be mocked as well. A way to to this would be that getRouteMock also mocks useRoute so that it returns the same object

Reply:
Yeah, it's kind of problematic that there is no global `navigation` object that can be easily accessed by testing code but rather it is available only inside components, which makes is hard to apply mocking using `spyOn`.